### PR TITLE
Add GINE capabilities to GIN

### PR DESCRIPTION
**Pull request by thegodone**
This is a tentative to implement GINE (adding edges)

Current code needs to define the edges size maybe this can be avoid.

## Proposed changes

tentative GINE example to address #141.

## Checklist

Put an `x` in the boxes that apply.
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] I have updated the necessary documentation

### CODE REVIEW

**Comment by francescofarina** [comment]
@thegodone it would be great to have this implemented within `mlx-graphs` rather than in a notebook.

**Comment by thegodone** [comment]
Of course, yes, but I am not sure of the result yet. can you first check my
code ?

Le dim. 10 mars 2024 à 10:36, Francesco Farina ***@***.***> a
écrit :

**Comment by TristanBilot** [comment]
For a next PR, it would be easier if you can write your code in real `.py` files instead of notebooks, as it's complicated to leave comments on a notebook.

Here are 
- Can you add `bias=True` in the `Linear` layer you instantiate? We generally follow PyG's implementation and they use bias.
- You use `edge_projection` instead of `edge_lin_proj`...
- `and not edge_features.shape[1] == node_features.shape[1]` -> you should apply the edge projection even if the node and edge shapes match.
- `hasattr(self, 'edge_projection')` -> you can use `'edge_projection' in self` instead
- your tests do not test GINE but only GINConv

**Comment by thegodone** [comment]
I update the python file gin_conv.py

File: mlx_graphs/nn/conv/gin_conv.py [hunk 1]
```diff
@@ -88,13 +111,19 @@ def __call__(
...
 
         aggr_features = self.propagate(
             edge_index=edge_index,
             node_features=node_features,
-            message_kwargs={"edge_weights": edge_weights},
+            message_kwargs={"edge_weights": edge_weights,
+                            "edge_features": edge_features},
         )
+        if 'edge_projection' in self:
+            aggr_features += edge_features
```
Comment (TristanBilot) [inline mlx_graphs/nn/conv/gin_conv.py]: You can't sum node embeddings with edge embeddings here, you have to do that in the `message` function.

**Comment by thegodone** [comment]
Thanks Tristan, for the next time I will follow PyG implementations. I would like to add the DMPNNconv what is the best way cause basically this is just MPNN but with a directed graph.

**Comment by TristanBilot** [comment]
I plan to merge this PR @thegodone, however you pushed a notebook related to the Esol dataset, can you remove it and cherry pick your work to a new branch for another PR?

**Comment by francescofarina** [comment]
> I plan to merge this PR @thegodone, however you pushed a notebook related to the Esol dataset, can you remove it and cherry pick your work to a new branch for another PR?

@TristanBilot I'll merge after your last review

**Pull request merged**